### ISSUE TITLE
Transfer Learning of Coco-trained model for new categories

### ISSUE THREAD

**Issue by TroubleMak3r** [fizyr/keras-retinanet#290]
Hi,

I'm trying to make a model localizing vehicle parts (eg. a bumper, windshield). Since my problem is very similar to localizing a car, (which is a very common object in Coco dataset and pre-trained model works remarkably well on it), I was thinking of loading a Coco pre-trained model, and re-train last layers on my own data (subset of Coco dataset), but with different labels. I've already manually labeled and localized a few houndred bumpers for starters, but I'm not sure what to do next... My idea is the following:

1. Load the coco pre-trained model (resnet50_coco_best...).
2. Set trainable parameter of the first N layers to False.
3. Create a custom data generator (using CSVGenerator).
4. Train the above model on a small subset of data with my custom generator.

However, I have following questions: 
1. Which layers to freeze ? The model has 215 layers, I was thinking of treezing the first  203, up to and including keras.engine.training.Model.
2. Will such model work on my custom dataset, with new labels and classes names (one to start with)? For starters, I get the 
`InvalidArgumentError (see above for traceback): Inputs to operation loss_1/classification_loss/Select_2 of type Select must have the same size and shape.  Input 0: [1,95193,1] != input 1: [1,95193,80]`
It's quite obvious since the original Coco dataset has 80 classes (and I have one - a bumper), but how to make it work on my model retaining the original weights and features up co certain layer?

Regads,
Dominik

**Comment by mxvs** [comment]
Hello Dominik,

Let be start by saying that you can't transfer learn the final classification set of layers due to the mismatch in the number of parameters, as you've already noticed. 

As long as you are on the correct version of Keras (2.1.3+) for Keras-Retinanet you can do something like this to steps to load the pre-trained weights:

```
#Create empty model with your number of classes
num_classes = 15
image = keras.layers.Input((None, None, 3))
model = resnet50_retinanet(num_classes, weights='imagenet', nms=True)

#Load weights, skipping incompatible sections
model.load_weights('resnet50_coco_best_v1.2.2.h5', skip_mismatch=True, by_name=True)
```
This should load the pre-trained weights for all layers except the `classification_submodel` due to the mismatch in the parameters caused by a difference in the number of classes you want to train on vs. the number of classes in the COCO dataset.

Then its down to experimentation to see what works for you dataset, you could try some of the following:

- Load pre-trained weights and train classification layer from scratch + fine-tune other retinanet layers (pyramid/regression) but freeze the backbone (resnet)
- Load pre-trained weights and only train classification layer from scratch while freezing all other layers
- Load pre-trained weights and unfreeze all retinanet layers + unfreeze the last convolutional block from the backbone (resnet), fine tune everything
- Don't load pre-trained weights (except for the resnet backbone imagenet weights) and train pyramid  regression + classification layers from scratch.
- Load pre-trained weights, first freeze everything except final classification layers and train those from scratch at a higher learning rate for a few epochs (to initialize them), then unfreeze additional (pre-trained) layers and continue training at a lower LR (to not unblance the existing pre-trained weights) and fine-tuning for a large(r) amount of epochs

There are probably quite a few more variations you could try and I would advise to explore them experimentally, potentially with different learning rates. 

PS. Another tip, use batch_size=1 at this time since a number of people have reported issues with higher batch sizes.

Hope this helps!

**Comment by mxvs** [comment]
No problem.

Just so you know, Keras-Retinanet has an Evaluation mechanism in the form of a callback baked in that you can use for testing purposes. See https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/callbacks/eval.py for more details.

**Comment by hgaiser** [comment]
Seems like @mxvs answered this nicely :) I'll close this then.
